### DISCUSSION TITLE
open webui multiple replicas

### DISCUSSION THREAD

**Discussion by ran-haim** [open-webui/open-webui#5109] [General]
I tried installing open webui using the helm chart.
It works fine, but when I setup more than one replica, it doesn't work.
One of the replicas always get 403, for example:
('127.0.0.6', 46874) - "WebSocket /ws/socket.io/?EIO=4&transport=websocket&sid=z1TTkpJJv6WiG0h-AAAs" 403
INFO:     connection rejected (403 Forbidden)

I also tried enabling external db, but it does not seems to be a way to do it using the values.yaml.

**Comment by tkg61** [comment]
i was also having this issue. Replicas set to 3, readwritemany pvc and no luck logging in. i figured it had to something with the fact that each request was being sent to a different server and i had seen other issues revolving around HA supposedly working for others. 

I assumed that i didn't need to do the external DB since the pvc was shared between the nodes but i will probably upgrade to that anyways later...

So i figured it was an ingress issue and did some digging. Found out that i should apply some sticky sessions to my ingress controller. Added these lines to the helm chart under the ingress annotations and it worked! 

nginx.ingress.kubernetes.io/affinity: "cookie"
nginx.ingress.kubernetes.io/session-cookie-name: "route"
nginx.ingress.kubernetes.io/session-cookie-expires: "172800"
nginx.ingress.kubernetes.io/session-cookie-max-age: "172800"

Also another good annotation to add is this one for larger file uploads

nginx.ingress.kubernetes.io/proxy-body-size: "4096m"

Anyways, i did some tests: 

- [X]  Log in from computer A, hit replica 0 and chat with model
- [X]  Log in from computer B, hit replica 1 and see my chat from computer A
    - [X] Chat in the same thread with computer B and refresh on computer A to see updates
- [X] Upload file from computer A, via replica 0 and do RAG with it
    - [X] Interact with the same thread with Computer B via replica 1 and continue to chat about the file
- [X] Upload a second file from Computer B, via replica 1 and chat about both files

So everything "seems" to work session wise. i haven't done a ton of tool and function testing yet but i will update if anything breaks.

**Reply by tkg61** [reply]
This isn't a perfect solution since the webui should be built to be completely stateless but i am not sure if it is there yet. @tjbck is this what we should be doing?

**Reply by lentiann** [reply]
I am having the same problem, and workaround was the same case.
In the helm charts deployment says if you have multiple replicas should use ReadWriteMany PVC, but did not work out also.
Maybe would he great if there's possible to use redis to manage session that would make central management.

**Reply by MrStrategy** [reply]
Thanks for this workaround. This actually helped me understanding the actual issue.

Currently I do use Nginx Ingress, OAuthProxy and OWUI and I try to scale all of them for HA and performance.
The sticky session approach is actually a bit tricky as I have the OAuth Proxy in between.
Do you have any good idea how I could tackle this issue?

**Reply by jk-f5** [reply]
As of #7887 with Redis and websocket support enabled, it shouldn't be necessary to configure sticky sessions for multiple replicas. I'd also recommend not using SQLite for multiple replicas.

**Reply by tkg61** [reply]
@jk-f5 Any chance you have a helm chart update to add redis along with some docs around proper multi-replicas? https://github.com/open-webui/helm-charts/tree/main

**Reply by tkg61** [reply]
> Thanks for this workaround. This actually helped me understanding the actual issue.
> 
> Currently I do use Nginx Ingress, OAuthProxy and OWUI and I try to scale all of them for HA and performance. The sticky session approach is actually a bit tricky as I have the OAuth Proxy in between. Do you have any good idea how I could tackle this issue?

oauth-proxy shouldn't acutally be in the way, it proxies oauth but still sits to the side and shouldn't be in the middle of traffic between nginx and OWUI

**Reply by jk-f5** [reply]
Sorry, I don't use the helm chart. I use terraform to deploy our k8s instances.

If you have oauth2-proxy setup and working with a single replica. Multiple replicas should work too. I assume you have `OAUTH2_PROXY_UPSTREAMS` set to the k8s service exposing your open webui k8s deployment?

**Reply by MrStrategy** [reply]
@jk-f5 Thanks for the tipps.
You are right, i have `OAUTH2_PROXY_UPSTREAMS` and was trying a back and forth with a single instance, multiple replicas etc.
What worked for me was bundling OAuth and OWUI into a single POD and using a local reference between these two. Then I could spin up multiple replicas and it works. 
But this is actually not a clean setup.

So I tried to setup websocket support (#7887), added a single valkey instance to my k8s cluster and made the necessary OWUI configuration. I can see in the debug logs it is actually using the `Using Redis to manage websockets.`
But now it looks like one of the instances works OK while the others don't. I do get a lot of ` 401 Unauthorized` on most of the services.
It`s a don't care if I have a single oauth2-proxy instance or multiple ones. Same outcome.

Any idea?

**Reply by jk-f5** [reply]
What does your ingress definition look like? Can you post the yaml?

**Reply by MrStrategy** [reply]
I do route the whole traffic through the OAuth2-proxy, so the config is rather simple:

Ingress:

apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: my-gpt-ingress
  namespace: prod    # Placeholder to be replaced
  annotations:
    nginx.ingress.kubernetes.io/rewrite-target: /
    nginx.ingress.kubernetes.io/ssl-redirect: "true"
    nginx.ingress.kubernetes.io/proxy-body-size: "50096m"

spec:
  ingressClassName: nginx
  tls:
  - hosts:
    - testgpt.my.com
    secretName: tls-my-cert
  rules:
  - host: testgpt.my.com
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: oauth2-proxy-service
            port:
              number: 4180


The controller is "standard":


apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-ingress-controller
  namespace: prod
spec:
  replicas: 2
  selector:
    matchLabels:
      app: nginx-ingress
  template:
    metadata:
      labels:
        app: nginx-ingress
    spec:
      serviceAccountName: nginx-ingress-serviceaccount
      containers:
      - name: nginx-ingress-controller
        image: registry.k8s.io/ingress-nginx/controller:v1.12.0
        args:
        - /nginx-ingress-controller
        env:
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        ports:
        - containerPort: 80
        - containerPort: 443


But what I can see even in a "single replica setup": When I open the url it gives a "not authorized" overwritten by "you are logged in now". So there may be something strange in my config.

edit: sorry for poor formatting but can't get it look better...

**Reply by jk-f5** [reply]
Judging by your ingress yaml, it looks like you have setup a variation of using oauth2-proxy as a reverse proxy with your nginx ingress sitting in front of oauth2-proxy (so you effectively have 2 reverse proxies sitting in front of open webui)
![image](https://github.com/user-attachments/assets/f2f03dd5-543a-4d58-9275-ee601a779ea2)

What I'd do here is configure your ingress to do the middleware route which requires setting the following annotations in your ingress definition. The response headers annotation is necessary only for setting whatever headers you need to identify the user to openwebui. Be sure that users must go through the ingress to get to openwebui because otherwise simply setting the required headers will allow access to openwebui. In other words, don't expose your openwebui service port externally otherwise all that is necessary for access is to set whatever header value you choose for the [trusted header](https://docs.openwebui.com/features/sso/#trusted-header) (be sure to read that link if you haven't)

```yaml
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: owui-ingress
  namespace: owui
  annotations:
    nginx.ingress.kubernetes.io/auth-url: 'http://oauth2-proxy.whatever/oauth2/auth'
    nginx.ingress.kubernetes.io/auth-signin: 'http://oauth2-proxy.whatever/oauth2/start'
    nginx.ingress.kubernetes.io/auth-response-headers: X-Auth-Request-Email
```

**Reply by MrStrategy** [reply]
Thanks @jk-f5,

i was already trying to migrate to the scenario you proposed. It took me one day to get the middleware working for a single replica scenario in a way you described...

Now I do no longer use the upstream of OAuth but the nginx ingress rules to handle the authentication. 
For the authentication itself we're using Microsoft AD and as people do have so many AD groups I also need to use redis in combination with oauth2 proxy.

This is how I understand my current scenario:

  - Unauthorized request gets forwarded to the auth provider for authenication. On successful auth oauth writes the auth cookie into redis and writes a browser cookie to represent the authentication
  - All following requests get forwarded to `oauth2/auth` url and the cookie gets decoded. OAuth injects the `X-Auth-Request-Email` information into the header
  - Open WebUI handles the `X-Auth-Request-Email` and and creates a cookie called `token` on its own valid for the current session.
  - With this in place you can use the application.
  
  Now I increased the replicas of the OpenWebUI pod from 1 to 2. (Ingress controller and OAuth still on 1 replica) and the system stops working again. Browser stays almost blank and in the logs of OWUI i can see a lot of unauthorized requests:
  
  
![image](https://github.com/user-attachments/assets/1d0637e6-7825-4f17-8215-39cb9e37b90f)

 
especially on the "second pod". So the pod i got authorized (which potentially generated the token) replies just with OK requests while the other one comes up with auth issues. 
I assume this is related to the token cookie but I'm not sure about it and i have no idea how to overcome this.

**Reply by MrStrategy** [reply]
OK, solved it on my own.

I did not set the `WEBUI_SECRET_KEY` variable. And while it defaults to a static value on non-docker installs it gets a random secret when running in a container. And looks like `WEBUI_SECRET_KEY` is used to encrypt the token. 
Now I set the value and all pods share the same secret and bang - it works.

Thank you so much for all your support.

**Reply by nwhobart** [reply]
I'm having a similar issue and I'm thankful for this thread; however the additional annotation on my `ingress-nginx` didn't quite do the trick. Currently my annotations look like this: 

```
      annotations:
        nginx.ingress.kubernetes.io/rewrite-target: /
        kubernetes.io/ingress.class: "nginx"
        nginx.ingress.kubernetes.io/whitelist-source-range: <secret  IPs>
        nginx.ingress.kubernetes.io/force-ssl-redirect: "true"
        nginx.ingress.kubernetes.io/proxy-body-size: 100m
        karpenter.sh/do-not-disrupt: "true"
        nginx.ingress.kubernetes.io/affinity: "cookie"
        nginx.ingress.kubernetes.io/session-cookie-name: "route"
        nginx.ingress.kubernetes.io/session-cookie-expires: "172800"
        nginx.ingress.kubernetes.io/session-cookie-max-age: "172800"
```

We are using okta for authentication. With a single replica there are no issues. As soon as a 2nd replica is introduced the log in just loops over and over again and I see this message in the logs:

```
 ERROR:    Exception in ASGI application
 Traceback (most recent call last):
   File "/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
     self.dialect.do_execute(
   File "/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 941, in do_execute
     cursor.execute(statement, parameters)
 sqlite3.OperationalError: no such table: group

sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: group
[SQL: SELECT "group".id AS group_id, "group".user_id AS group_user_id, "group".name AS group_name, "group".description AS group_description, "group".data AS group_data, "group
FROM "group"
WHERE json_array_length("group".user_ids) > ? AND CAST("group".user_ids AS VARCHAR) LIKE ? ORDER BY "group".updated_at DESC]
[parameters: (0, '%"21d82c5b-6496-40ea-a02d-6d395e9e0e6d"%')]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
```

Any help would be appreciated. We are using the standard Helm chart.

**Reply by jk-f5** [reply]
It looks like your replica isn't using the same database as the other. I would highly recommend **not** using Sqlite as your application database and use Postgres instead. Otherwise you are going to have to share a persistent volume between your replicas and they'll all be accessing the same backing sqlite database file. 

If you have your replicas, postgres, and redis configured correctly. These annotations below are not necessary:
```javascript
        nginx.ingress.kubernetes.io/affinity: "cookie"
        nginx.ingress.kubernetes.io/session-cookie-name: "route"
        nginx.ingress.kubernetes.io/session-cookie-expires: "172800"
        nginx.ingress.kubernetes.io/session-cookie-max-age: "172800"
```

**Reply by nwhobart** [reply]
Thanks @jk-f5 ... Is there thorough documentation on this? Running multiple replicas seems to be something that should be documented better unless I'm just missing something. 

I thought using the EFS `storageclass` using `ReadWriteMany` would take care of all that business.

**Reply by tjbck** [reply]
@nwhobart would genuinely appreciate a community contribution for documenting the process here!

**Reply by nwhobart** [reply]
@jk-f5 @tjbck I'm now on a postgres db in aws rds and i still get auth errors with multiple replicas. is redis required for multiple replicas? 

```
WARNI [open_webui.utils.oauth] OAuth callback error: mismatching_state: CSRF Warning! State not equal in request and response.
```

the ui shows: 

```
{"detail":"The email or password provided is incorrect. Please check for typos and try logging in again."}
```

On a single replica the oauth portion works just fine.

**Comment by rragundez** [comment]
@jk-f5 just to clarify, are you saying that it is enough to set this variables in each replica plus having an external Postgredb?

ENABLE_WEBSOCKET_SUPPORT="True"
WEBSOCKET_MANAGER="redis"
WEBSOCKET_REDIS_URL="redis://redis:6379/1"

**Reply by jk-f5** [reply]
Yes, assuming you have also configured a Redis instance that has a host name of `redis` that correctly resolves via DNS in your environment. (I assume you already have a Redis database on `/0` since you chose `/1` in your Redis URL example above.)

[Here is how I run Open Webui in a dev environment that sets up the whole stack in docker compose.](https://github.com/jk-f5/open-webui-dev-env)

**Reply by rragundez** [reply]
thanks, it works.

**Reply by rragundez** [reply]
@jk-f5  I just enabled image generation and on this part is not working. When an image is generated it is stored into disk and retrieved by an API request by the frontend, if the request is not sent to the same machine then it gives 404. 
This is within the same conversation, even within the same message exchange of generating the image, I see the image is generated by is not displayed when trying to retrieve it. 
What do you think is going on with this one? something I might need to tweak in the config or ...? thanks in advance

**Reply by jk-f5** [reply]
Good callout, you will have to mount a PVC to /app/backend/data directory to make the `cache` and `uploads` directory available between replicas. Bummer.

**Reply by rragundez** [reply]
it does work with uploaded files though, but that I think is because the metadata info is stored and retrieved from the DB. Couldn't we do something similar?

Also If I am not mistaken the retrieval of the images there is no check on the user ID, any user could potentially retieve any other user's generated images.

**Reply by jk-f5** [reply]
Definitely agree that we need to eliminate the need for a volume mount.

Assuming your second comment is true, you might want to create an issue specifically for that.

**Reply by mmcaulay** [reply]
@rragundez would that have anything to do with permissions checks needing a tweak as in #8889 & #8483? it might not be multiple replicas

**Reply by rragundez** [reply]
I don't think so, is just how the images are served. They are stored in the cached/images/generations directory which is public and served directly from there in img html components.

We change the source code to use a GCS bucket as intermediary and add it a middleware endpoint to check authentication and served the images from there. 
Ideally I think it should follow the same of the files, where the metadata is saved to the DB

**Reply by rragundez** [reply]
I created a PR to solve the issue: 
https://github.com/open-webui/open-webui/pull/9486

**Comment by tkg61** [comment]
Ok... interesting results with this. i hope someone can replicate this... 


### I have gotten 3 replicas working with .5.4 

My config: 
- K8s 1.30
- Latest open-webui helm chart (5.4.0)
- Websockets turned on
- Working redis with debug log confirming
- Working single node postgres shared between openwebui replicas
- Shared openwebui secret key
- Only nginx annotation i have is: 
   - nginx.ingress.kubernetes.io/proxy-body-size: 4096m

ENV: 
```
CONTENT_EXTRACTION_ENGINE : Tika
DATABASE_URL : postgresql://postgres:PASSWORD@psql-test-postgresql:5432/postgres
ENABLE_OLLAMA_API : False
ENABLE_WEBSOCKET_SUPPORT : True
OPENAI_API_BASE_URL : http://open-webui-pipelines.owui-test.svc.cluster.local:9099
SOCKET_LOG_LEVEL : DEBUG
TIKA_SERVER_URL : http://open-webui-tika:9998
WEBSOCKET_MANAGER : redis
WEBSOCKET_REDIS_URL : redis://open-webui-redis-master:6379/1
WEBUI_SECRET_KEY : tQD5RHiU42ubYJ1SeRn1
```

Versions: .5.5, .5.6, .5.7

Now...I have tested with these and still am not having luck. 1 replica works just fine but anymore still breaks. Port-forwarding around nginx directly to the service works fine too. 

   
Chrome console is printing: 

```
Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "text/html". Strict MIME type checking is enforced for module scripts per HTML spec.Understand this errorAI
entry.DtRrzbjw.js:1 Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "text/html". Strict MIME type checking is enforced for module scripts per HTML spec.Understand this errorAI
index.CMgO2_ru.js:1 Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "text/html". Strict MIME type checking is enforced for module scripts per HTML spec.Understand this errorAI
70954b3c-1405-4f9b-bdc8-826e9259fa27:1 Uncaught (in promise) TypeError: Failed to fetch dynamically imported module: 
```

**Reply by jk-f5** [reply]
What does the network tab of your browser's developer tools show? You should be able to see and confirm that those javascript libraries are being pulled and their mime types.

Are you sure you don't have some javascript cached from an older version of open webui. Try clearing your browser cache and refreshing open webui.

**Reply by tkg61** [reply]
Yes, we have tested 3 different machines, varying chrome and MS Edge versions (including latest) with cleared caches and incognito modes as well. 

here is a screenshot of the network tab and console output from a fresh browser in incognito

![image](https://github.com/user-attachments/assets/45f7c220-9c92-409e-a956-249d333e131f)

**Reply by jk-f5** [reply]
If it's returning a mime type of text/html with an http status of 200, I'd be interested to see what the content of these responses actually is.

Check the http request and response headers, then check the actual response itself and see if it actually is returning json with a mime type of text/html. 

Example:
![image](https://github.com/user-attachments/assets/b7eb5951-c997-49bc-9601-86bbace070d6)
![image](https://github.com/user-attachments/assets/c7818078-1e0b-42a9-93d9-f80e7ec14e3e)

**Reply by tkg61** [reply]
Here's what i see, thanks for taking a look! 

Headers are HTML: 

![image](https://github.com/user-attachments/assets/b9a4991d-df16-44c3-829f-a59a33e10ef6)

and it is returning HTML: 

![image](https://github.com/user-attachments/assets/0f11bea9-5257-4cdf-811b-574b717352ec)

**Reply by jk-f5** [reply]
That is the contents of App.html which suggests that the fallback route for the single‐page app is intercepting requests for the JS file (because FastAPI cannot find a matching static route). In other words, FastAPI is failing to serve the actual .js file from a static directory and is instead returning App.html

Are you running open webui with any persistent volumes attached? If so, what are they?

Are you absolutely sure there isn't another replica running a different version of open webui?

**Reply by tkg61** [reply]
Hmmm i will try some more tests in the morning since i am curious why this works with .5.4 but not the newer versions but here are the pods that are running in this namespace. i have only been bouncing between versions one at a time and have uninstalled helm chart when switching but i am wondering if the postgres server needs to be wiped as well between versions, just sort of surprised by that. 

Here is the persistence part of the helm chart: 

```
persistence:
  accessModes:
  - ReadWriteMany
  annotations: {}
  enabled: true
  existingClaim: ""
  selector: {}
  size: 2Gi
  storageClass: longhorn
  subPath: ""
```

Pods: 

![image](https://github.com/user-attachments/assets/fb686b9d-06ee-429c-8e87-645af49945bf)

i think my next test will be wiping the postgres and starting with .5.7 instead of upgrading to it but not being able to upgrade would be confusing and i will re-test that again as well

**Reply by jk-f5** [reply]
FYI: You don't need any persistent volume at all for the open webui k8s deployment if you're using Postgres.

I have been upgrading open webui through many releases in k8s with no problem. We just need to figure out what's isn't configured correctly on your end.

I don't run the helm chart personally (I have opinions about helm) but if you post your values.yaml (with anything sensitive replaced) and what version of the open webui helm chart you are running, I'll take a look.

**Reply by tkg61** [reply]
edit: added ingress

So i think according to this ticket, since we start off before the S3 feature, we still have to have persistence for our uploaded files since there isn't a way (that i know of) to move them to S3: https://github.com/open-webui/helm-charts/issues/83 but that is good to know otherwise. 

i have tested again with a .5.7 cluster from the start with a new namespace/ingress so everything is fresh and still have the MIME issue. If i downgrade to .5.4 from .5.7 it works immediately.

Install process: 

- `helm install psql-test bitnami/postgresql --namespace owui-test`
- Get password from configmap and place in values file
- `helm upgrade --install open-webui open-webui/open-webui --values .\new-owui-values.yaml --namespace owui-test --set 'extraEnvVars[0].name=SOCKET_LOG_LEVEL' --set 'extraEnvVars[0].value=DEBUG'`
- Add ingress: 
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: owui-test
  annotations:
    nginx.ingress.kubernetes.io/proxy-body-size: 4096m
spec:
  ingressClassName: nginx
  rules:
    - host: owui-test.mydomain.com
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: open-webui
                port:
                  number: 80
```
Here is the helm values file attached, will want to rename to .yaml
[new-owui-values - Copy.txt](https://github.com/user-attachments/files/18623452/new-owui-values.-.Copy.txt)

**Reply by jk-f5** [reply]
What you have above works for me.

I did a fresh install using [KinD 1.32.1](https://kind.sigs.k8s.io/) with exactly what you have above except I first altered your `values.yaml` to use `0.5.4`. I verified it was behaving with `0.5.4`, then I modified your `values.yaml` to change `0.5.4` to `0.5.7` and reran the `helm upgrade --install` again. Works fine after the upgrade. 

If you're interested, I can post exactly what I ran to create my environment from scratch.

**Reply by tkg61** [reply]
This was helpful, i am doing more testing and have made some progress where i believe i am able to replicate what you did in KinD and have it work. I am now trying to diagnose our environmental problems as to what might be the issue. Will keep this updated

**Reply by justbert** [reply]
Hey @tkg61,
I moved over from local uploads to s3 today. All you need to do is get your s3 setup, copy over the documents from the upload folder to the s3 instance, then update the path for the files in the `file` table:
`UPDATE file SET path = REPLACE(path,'/app/backend/data/uploads/','s3://$BUCKET_NAME/') where  path like '/app%';`

To copy over the docs, I mounted the volume to a pod with an s3 CLI. You can usually mount the ReadWriteOnce volume at the same time as your openwebui instance if they're running on the same node. I used a `podAffinity` to get them scheduled together.

**Reply by cruzanstx** [reply]
This was super helpful. WEBUI_SECRET_KEY was my issue, each instance had a different key because I didn't set it as an env variable in my helm chart.

Kubernetes Version: v1.25.10 +k3s1
CONTENT_EXTRACTION_ENGINE : Tika
DATABASE_URL : postgresql://postgres:PASSWORD@<internal_postgres>:5432/postgres
ENABLE_OLLAMA_API : False
ENABLE_WEBSOCKET_SUPPORT : True
OPENAI_API_BASE_URL : http://<internal_pipelines>:9099
SOCKET_LOG_LEVEL : DEBUG
TIKA_SERVER_URL : http://<internal_tika>:9998
WEBSOCKET_MANAGER : redis
WEBSOCKET_REDIS_URL : redis://<internal_redis>:6379/1
WEBUI_SECRET_KEY : <KEY>

**Comment by Abdelrahman1993** [comment]
Hi All
May I know why Redis is needed for the OpenwebUI!!
Is it used as a session management in multi replica deployment ?

**Reply by jk-f5** [reply]
socket.io (websocket) session management in a multi replica deployment specifically

**Reply by Abdelrahman1993** [reply]
It is not needed in one instance deployment, right ?
It is needed only in multi instance deployment to share the session in central location, am i understand it right ?

**Reply by jk-f5** [reply]
That's correct. Though you could still use Redis in the single replica setup as well.

**Reply by Abdelrahman1993** [reply]
Thank you 👍

**Comment by rragundez** [comment]
@jk-f5   I am seeing an error when trying to stop the stream of a response from the model. I am pretty sure it is also related with multiple replicas. 
For example is calling: 
`/api/tasks/stop/8c97871a-c9cc-4087-bf1b-0e1296796570`
but then I get 404
![image](https://github.com/user-attachments/assets/366a1b79-1858-42e3-9252-b9934d0c4b74)

when I look at the stop_tasks mechanism, 
https://github.com/open-webui/open-webui/blob/main/backend/open_webui/tasks.py#L49
https://github.com/open-webui/open-webui/blob/main/backend/open_webui/main.py#L950

I am pretty sure is because the stop request lands in another replica where the tasks dictionary doesn't know about that task id. 

Any idea how would we handle that using the REDIS DB or any other way?

**Reply by ips972** [reply]
Hi all. Did the above. Expanded my setup to 3 open webui docker nodes behind a load balancer. (HAPROXY) 
Enabled the socket handling and setup a redis node in the same compose.yml file. Everything comes up fine. No error messages. If I start on node1 and take it. Down. I’m transferred to node 2 and 3.. etc.. with stickiness by cookie. 

this strange thing is, even though I had the env commands for this in all 3 services in the docker-compose.yml file. I couldn’t see any traffic between any of the open webui nodes to the redis in tcpdump -nni any port 6379
How can I verify that redis is infact being used? 

Thanks!

**Comment by roeizavida** [comment]
Was anyone able to use multiple replicas in Kubernetes? It seems like the replicas are not in sync even though they are connected to the same PostgreSQL database and Redis instance.
For example, when I disable an OpenAI connection, I only see it when reaching one of the replicas and for the rest, I see that it is still enabled. It's the same in the models tab. Sometimes I see the models from that connection, and sometimes not.
The model responded in the chat also seem slower when using multiple replicas.
Reducing to a single replica solves it, but then it is not highly available.

**Reply by jyje** [reply]
@roeizavida Hello, I deployed open-webui with HA using https://github.com/open-webui/helm-charts .

following helm values can help you to make it.
```yaml
persistence:
  enabled: true
replicaCount: 3
websocket:
  enabled: true
ingress:
  enabled: true
  class: nginx
  annotations:
    nginx.ingress.kubernetes.io/session-cookie-name: "route"
    nginx.ingress.kubernetes.io/session-cookie-expires: "172800"
    nginx.ingress.kubernetes.io/session-cookie-max-age: "172800"
```

you can see full [example values in my cluster](https://github.com/jyje/cluster/blob/main/clusters/r4spi/apps/open-webui.yaml) in ArgoCD CRD. Please let me know how you tried!

**Reply by tkg61** [reply]
I think you are also running into this issue: #10365 where multiple replicas don’t sync settings.

in the end I actually had to switch clusters then I was able to remove the nginx annotations to have the multiple replicas work. Never figured out why since it was taking too long to figure out

Edit: Oh also make sure you are using the same secret key talked about in this thread

**Reply by roeizavida** [reply]
@jyje Thanks! Seems like using Nginx + the annotations you provided solved this.

**Reply by arnonearth** [reply]
@roeizavida Is it because you set the sticky session so the one who made configuration change is routed to the same instance that has been updated but other replicas still do not get the synced? I have followed @tkg61 issue and found out that I only set WEBSOCKET_REDIS_URL. Now I have both WEBSOCKET_REDIS_URL and REDIS_URL configured, the configs seem to share between each replica now.

**Comment by mahenning** [comment]
Maybe someone with a multi-replica setup can help me. I have the pull request https://github.com/open-webui/open-webui/pull/13225 for re-indexing files with UI feedback/loading bar and need multi-replica testing because I only have a single instance. It would be super helpful if someone with multi-replicas could run the PR version of the code and test if the loading bar/reindexing stuff is working there, or even just review the code with the redis stuff. Feedback in the comments of the PR is much appreciated. Thanks!
